### PR TITLE
[server] Make sure to only ever finish this span once

### LINKED ISSUE #16228: Error: You can only call finish() on a span once.
Logs contain errors with span finishing

https://console.cloud.google.com/logs/query;query=resource.labels.pod_name%20:%20%2528%22server%22%2529%0A%22call%20finish%2528%2529%20on%20a%20span%20once%22;timeRange=PT1H;cursorTimestamp=2023-02-06T12:43:34.616Z?authuser=0&project=gitpod-191109

### PR DESCRIPTION
**Pull request by geropl**
## Related Issue(s)

Fixes #16228

## Release Notes

```release-note
NONE
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all
- [ ] /werft no-test
      Run Leeway with `--dont-test`
- [ ] /werft publish-to-npm

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/gitpod-protocol/src/util/tracing.ts [hunk 1]
```diff
@@ -186,6 +186,15 @@ export namespace TraceContext {
             context: owi,
         });
     }
+
+    export function finishOnce(span: opentracing.Span): () => void {
```
Comment (easyCZ) [inline components/gitpod-protocol/src/util/tracing.ts:190]: Sounds like a hack on top of improper control flow.. but it will do

**Review by easyCZ — APPROVED** [review]

File: components/gitpod-protocol/src/util/tracing.ts (hunk 1 above)
Comment (geropl) [reply components/gitpod-protocol/src/util/tracing.ts:190]: Sometimes it's hard to know all the ways sth can behave, like in this case. Then the solution becomes a trade-off about where we want to spent our time. :shrug:

**Comment by geropl** [comment]
@WVerlaek I just noticed @csweichel got pulled in for a change on `image-builder-api`. Would it make sense to switch ownership to you? :thinking:

File: components/gitpod-protocol/src/util/tracing.ts (hunk 1 above)
Comment (easyCZ) [reply components/gitpod-protocol/src/util/tracing.ts:190]: Well, the span should always be opened on a middleware, and always closed on the middleware, irrespective of the outcome of individual functions.

In a way, we shouldn't even have to create the span at the start of the method, just add to a span which exists.

Comment (geropl) [reply components/gitpod-protocol/src/util/tracing.ts:190]: >  the span should always be opened on a middleware, and always closed on the middleware

That's certainly sth we can aim for in the future. :+1: 

> In a way, we shouldn't even have to create the span at the start of the method, just add to a span which exists.

That's a question on granularity, which would also be interesting to discuss and align on. :100:

Comment (easyCZ) [reply components/gitpod-protocol/src/util/tracing.ts:190]: This is one aspect where a service oriented architecture helps. Once the tracing is setup once, you get the same behaviour across all (baseserver) servers, and you get normalized tracing, even when you've not instrumented it explicitly. You can always add more child spans, when relevant.

Propagation of these, through context ends up being natural in Go.

**Comment by WVerlaek** [comment]
@geropl good point, maybe better to extend ownership wider to team-workspace as well, in that case I wouldn't be blocking on reviews to the image builder api? Wdyt @kylos101 @csweichel ?

**Comment by kylos101** [comment]
@WVerlaek api's are generally limited to small lists of individuals, rather than teams, at least with components owned by Team Workspace. Given that, and your recent image-builder work, your're a good fit. :smiley:  At least, as a short term solution, to help with reviewing this PR. Can you submit a separate PR to change the owner?

Long term, given our new team size, I think this strategy is worth revisiting. We may want to consider swapping a few APIs team ownership. Could you add as a discussion point for our next team sync?

**Review by WVerlaek — APPROVED** [review]

**Pull request merged**